### PR TITLE
Introduce private and unstable plugin traits

### PR DESCRIPTION
**Pull request by Geal**
The goal of those traits is to allow experimentation and development on new plugin hook points without committing to a public API right away.

We add to new traits:
- PluginUnstable: contains experimental plugin methods that can be used publicly but for which there's no committment to stability, they can change at any moment
- PluginPrivate: internal trait that is not exposed publicly

This implements the decision record introduced in https://github.com/apollographql/router/commit/43c56400302ee754944866a092aef96957e0e480

*Description here*

Fixes #**issue_number**

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugin/mod.rs [hunk 1]
```diff
@@ -360,13 +626,19 @@ pub(crate) trait DynPlugin: Send + Sync + 'static {
 
     /// Support downcasting
     fn as_any_mut(&mut self) -> &mut dyn std::any::Any;
+
+    /// unstable
+    fn unstable_method(&self);
+
+    /// internal
+    fn internal_method(&self, i: InternalType);
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/plugin/mod.rs:662]: (I'll remove them before merging)

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/test_harness.rs [hunk 2]
```diff
@@ -165,6 +167,45 @@ impl<'a> TestHarness<'a> {
         self
     }
 
+    /// Adds an extra, already instanciated unstable plugin.
```
Comment (garypen) [inline apollo-router/src/test_harness.rs:170]:
```suggestion
    /// Adds an extra, already instantiated unstable plugin.
```

File: apollo-router/src/test_harness.rs [hunk 3]
```diff
@@ -165,6 +167,45 @@ impl<'a> TestHarness<'a> {
...
+                self.extra_plugins.len(),
+                std::any::type_name::<P>(),
+            ),
+        };
+
+        self.extra_plugins.push((name, Box::new(plugin)));
+        self
+    }
+
+    /// Adds an extra, already instanciated private plugin.
```
Comment (garypen) [inline apollo-router/src/test_harness.rs:189]:
```suggestion
    /// Adds an extra, already instantiated private plugin.
```

File: apollo-router/src/plugin/mod.rs [hunk 4]
```diff
@@ -437,6 +747,42 @@ macro_rules! register_plugin {
     };
 }
 
+/// Register a private plugin with a group and a name
+/// Grouping prevent name clashes for plugins, so choose something unique, like your domain name.
+/// Plugins will appear in the configuration as a layer property called: {group}.{name}
+#[macro_export]
+macro_rules! register_private_plugin {
```
Comment (garypen) [inline apollo-router/src/plugin/mod.rs:754]: Are we missing `register_unstable_plugin` ?

File: apollo-router/src/plugin/mod.rs [hunk 5]
```diff
@@ -187,8 +187,36 @@ impl PluginFactory {
     pub(crate) fn is_apollo(&self) -> bool {
         self.name.starts_with("apollo.") || self.name.starts_with("experimental.")
     }
+
+    /// Create a plugin factory.
+    pub fn new<P: PluginUnstable>(group: &str, name: &str) -> PluginFactory {
```
Comment (garypen) [inline apollo-router/src/plugin/mod.rs:192]: Perhaps better named as `new_unstable` ?

File: dev-docs/decision-records/unstable-plugins.md [hunk 6]
```diff
@@ -23,14 +23,13 @@ Part of the motivation for this decision, was the recognition that we already ha
```
Comment (garypen) [inline dev-docs/decision-records/unstable-plugins.md:23]: The changes to this document don't seem consistent with the changes to the code. For instance, the code has a `PluginPrivate`, but it's removed from the DR. Are these changes not properly synced up?

**Review by garypen — COMMENTED** [review]
Looks good, but I can't reconcile the code changes with the DR changes.

File: dev-docs/decision-records/unstable-plugins.md (hunk 6 above)
Comment (BrynCooke) [reply dev-docs/decision-records/unstable-plugins.md:23]: Suspect this file just needs to be reverted

File: apollo-router/src/plugin/mod.rs (hunk 5 above)
Comment (BrynCooke) [reply apollo-router/src/plugin/mod.rs:192]: What's the reason we need `new` and `new_private` at all?

Comment (Geal) [reply apollo-router/src/plugin/mod.rs:192]: it is used in the `register_plugin` macro

File: dev-docs/decision-records/unstable-plugins.md (hunk 6 above)
Comment (Geal) [reply dev-docs/decision-records/unstable-plugins.md]: it's reverted now

**Comment by Geal** [comment]
merging this without changelog entries, as the point of those traits is internal work

**Pull request merged**